drivers: riscv: Add AIA (Advanced Interrupt Architecture) support

AFOliveira opened the pull request:
This PR implements support for RISC-V Advanced Interrupt Architecture (AIA) as proposed in RFC #82487, with some changes from the RFC scope.

## Overview

Implements the APLIC (Advanced Platform-Level Interrupt Controller) and IMSIC (Incoming Message-Signaled Interrupt Controller) drivers, enabling AIA support for RISC-V SoCs. The implementation provides a PLIC-compatible interface while supporting modern MSI-based interrupt delivery. Also enables QEMU support to ensure correctness.

## Implementation

Following the RFC guidance, this implementation:

- **Presents a PLIC-like interface**: Applications use standard `irq_enable()` / `irq_disable()` APIs
- **Hides AIA complexity**: Unified `riscv_aia.h` API abstracts APLIC+IMSIC hierarchy and MSI configuration
- **Integrates with privileged RISC-V common code**: Uses same pattern as PLIC in `soc_common_irq.c`
- **Supports wired interrupts**: Full peripheral interrupt routing via APLIC → IMSIC → MEXT

## Beyond RFC Scope

The RFC proposed a single driver that "could be subdivided internally" and would later be "reworked into two separate drivers." This implementation goes directly to that split architecture:

- **Separate drivers from start**: `intc_riscv_imsic.c`, `intc_riscv_aplic_msi.c`, and `intc_riscv_aia.c` coordinator
- **SMP support**: Per-CPU IMSIC initialization with interrupt migration between harts
- **Devicetree separation**: Separate nodes for APLIC and per-hart IMSIC instances

## Testing

- Two working samples: `uart_echo_aia` (single-hart) and `uart_echo_aia_smp` (multi-hart)
- Test suite
- Tested on QEMU RISC-V32 AIA platform
- Demonstrates interrupt delivery, routing, and hart migration


I am keeping this as draft for now as I still need to catch the _infinite_ loose ends that continuously appear. Reviews are very welcome.

fkokosinski commented on dts/bindings/interrupt-controller/riscv,imsic.yaml:
```diff
@@ -0,0 +1,26 @@
+  riscv,hart-id:
+    type: int
+    description: Hart ID associated with this IMSIC instance
+    default: 0
```
ditto

fkokosinski commented on drivers/interrupt_controller/intc_riscv_imsic.c:
```diff
@@ -0,0 +1,244 @@
+static void imsic_irq_config_func_0(void)
+{
+	/* Only hart 0 (instance 0) registers the global MEXT IRQ handler */
+	IRQ_CONNECT(11, 0, imsic_mext_isr, DEVICE_DT_INST_GET(0), 0);
```
Hardcoded IRQ number?

AFOliveira replied:
It's spec-mandated that interrupt 11 is Machine External Interrupt, however I think that magic wasn't very smart on my side, do you agree with substituting it with RISCV_IRQ_MEXT macro?

fkokosinski commented on include/zephyr/drivers/interrupt_controller/riscv_imsic.h:
```diff
@@ -89,10 +94,12 @@ void riscv_imsic_complete(uint32_t eiid);
- * @return 0 on success, negative errno on failure
  */
-int riscv_imsic_enable_eiid(uint32_t eiid);
+void riscv_imsic_enable_eiid(uint32_t eiid);
```
shouldn't these symbols be declared as voids in the previous commit?

AFOliveira replied:
of course, I just realized I messed commit history, I'll fix it.

fkokosinski commented on include/zephyr/drivers/interrupt_controller/riscv_imsic.h:
```diff
@@ -101,10 +108,12 @@ int riscv_imsic_enable_eiid(uint32_t eiid);
- * @return 0 on success, negative errno on failure
  */
-int riscv_imsic_disable_eiid(uint32_t eiid);
+void riscv_imsic_disable_eiid(uint32_t eiid);
```
ditto

fkokosinski reviewed (commented):
Thanks a lot for this PR! Please note that since the changes you introduce here are significant (2,5k+ LOC added), it will take some time for me to dig through this :)

Here's some remarks/questions for starters

fkokosinski reviewed (commented):
Some problems w/r to the commit history still unfortunately persist.

Could we split this PR into two: one which adds the base AIA support and one of the irq controller drivers, and then a subsequent one adding the other controller?

Judging by Kconfig, these two are fairly independent of each other. I reckon that would reduce the review time quite significantly.

AFOliveira commented on drivers/interrupt_controller/intc_riscv_imsic.c:
```diff
@@ -0,0 +1,247 @@
+/*
```
I tried to put them both (APLIC and IMSIC) inside the AIA Kconfig as AIA to keep them unified, i.e. I don't expect anyone to use both drivers separately.

AFOliveira committed:
include: drivers: interrupt_controller: riscv: add APLIC header

Add public APLIC-MSI driver API header for source configuration and MSI
routing.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
arch: riscv: add RISCV_HAS_AIA capability

Add a capability symbol for platforms providing RISC-V AIA interrupt
controllers.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
riscv: integrate AIA interrupt management

Integrate AIA handling into common RISC-V interrupt enable/disable and
priority setup.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
include: drivers: interrupt_controller: riscv: add AIA header

Add public header for the RISC-V AIA coordinator API.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
dts: bindings: riscv: add APLIC-MSI binding

Add devicetree binding for the RISC-V APLIC in MSI delivery mode.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
tests: drivers: interrupt_controller: add RISC-V AIA tests

Add tests covering IMSIC delivery and APLIC-MSI routing on QEMU AIA
platforms.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
samples: drivers: interrupt_controller: add RISC-V AIA SMP demo

Add SMP demonstration for RISC-V AIA showing multi-hart interrupt
handling with per-hart IMSIC initialization and APLIC MSI routing.

Features demonstrated:
- Multi-hart boot with per-hart IMSIC initialization
- APLIC MSI routing to specific harts
- Software MSI injection (GENMSI) targeting individual harts

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
samples: drivers: interrupt_controller: add RISC-V AIA demo

Add a platform-agnostic AIA demonstration sample that shows the
interrupt flow: UART → APLIC → MSI → IMSIC → CPU.

The sample configures APLIC source routing, routes UART interrupts
via MSI to IMSIC, and echoes typed characters.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
tests: add qemu_riscv32 AIA board variants

Update tests and overlays to cover qemu_riscv32 AIA board variants.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
boards: qemu: riscv32: add AIA board variants

Add qemu_riscv32 board qualifiers for AIA single-core and SMP variants.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
dts: qemu: riscv: refactor PLIC interrupt controller placement

Move PLIC definition from base virt-riscv.dtsi to individual board files
that actually use PLIC. This eliminates inheritance of PLIC by AIA boards
that don't need it.

Changes:
- Remove PLIC from dts/riscv/qemu/virt-riscv.dtsi
- Add PLIC to boards/qemu/riscv32e/qemu_riscv32e.dts
- Add PLIC to boards/qemu/riscv64/qemu_riscv64.dts
- Add PLIC to boards/qemu/riscv64/qemu_riscv64_qemu_virt_riscv64_smp.dts

This makes the device tree structure cleaner and allows AIA boards to
define their own interrupt controllers without needing /delete-node/
workarounds.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
soc: qemu: riscv: add AIA-enabled RV32 SoC variant

Add qemu_virt_riscv32_aia SoC variant with AIA interrupt controller
support.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
boards: qemu: riscv32: use explicit SoC qualifiers

Switch qemu_riscv32 references to explicit SoC-qualified platform
identifiers.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
tests/samples: update qemu_riscv32 platform identifiers

Update test/sample metadata to match qemu_riscv32 multi-SoC platform
identifiers.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
drivers: interrupt_controller: riscv: add AIA coordinator

Add coordinator driver for RISC-V Advanced Interrupt Architecture (AIA)
that provides a unified API wrapping APLIC and IMSIC operations.

The coordinator simplifies application code by providing wrapper functions
for common AIA operations:
- IRQ enable/disable: Manages IMSIC EIID enable/disable
- Source configuration: Configures APLIC source trigger modes
- MSI routing: Routes APLIC sources to specific harts and EIIDs
- Source enable: Enables APLIC interrupt sources
- GENMSI injection: Software-triggered MSI via APLIC GENMSI

Note: In AIA-MSI mode, APLIC sources and IMSIC EIIDs are decoupled.
irq_enable() only enables the EIID in IMSIC; APLIC sources must be
configured separately via the source configuration APIs.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
drivers: interrupt_controller: riscv: add APLIC-MSI Kconfig

Add Kconfig symbol for RISC-V APLIC in MSI delivery mode.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
drivers: interrupt_controller: riscv: add APLIC-MSI driver

Add the APLIC-MSI driver implementation that routes wired interrupts
to the IMSIC as Message-Signaled Interrupts.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
arch: riscv: rename icsr_* to micsr_* for M-mode indirect CSR access

Rename the indirect CSR access functions from icsr_* to micsr_* to
explicitly indicate they operate using M-mode CSRs (MISELECT/MIREG).

This prepares for future S-mode support where sicsr_* functions using
SISELECT/SIREG would be needed. The naming follows the RISC-V
convention of using 'm' and 's' prefixes to distinguish privilege
levels (e.g., mstatus/sstatus, mie/sie).

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
drivers: interrupt_controller: riscv: imsic: use micsr_* functions

Update the IMSIC driver to use the renamed micsr_* indirect CSR
access functions instead of the old icsr_* names.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>
